driver-core/src/main/java/com/datastax/driver/core/QueryLogger.java

```diff
@@ -0,0 +1,544 @@
+
+    /**
+     * The logger used to log slow queries, i.e., queries that completed successfully
+     * in more than {@link #DEFAULT_SLOW_QUERY_THRESHOLD_MS} milliseconds.
+     * This logger is activated by setting its level to {@code DEBUG} or {@code TRACE}.
+     * Additionally, if the level is set to {@code TRACE} and the statement being logged is a {@link BoundStatement},
+     * then the query parameters (if any) will be logged as well (names and actual values).
+     */
+    public static final Logger SLOW_LOGGER = LoggerFactory.getLogger("com.datastax.driver.core.QueryLogger.SLOW");
+
+    /**
+     * The logger used to log unsuccessful queries, i.e., queries that did not completed normally and threw an exception.
+     * This logger is activated by setting its level to {@code DEBUG} or {@code TRACE}.
+     * Additionally, if the level is set to {@code TRACE} and the statement being logged is a {@link BoundStatement},
+     * then the query parameters (if any) will be logged as well (names and actual values).
+     * Note this this logger will also print the full stack trace of the reported exception.
+     */
+    public static final Logger ERROR_LOGGER = LoggerFactory.getLogger("com.datastax.driver.core.QueryLogger.ERROR");
+
+    // Message templates
+
+    private static final String NORMAL_TEMPLATE = "Query completed normally on host %s, took %s ms: %s";
+
+    private static final String SLOW_TEMPLATE = "Query too slow on host %s, took %s ms: %s";
+
+    private static final String ERROR_TEMPLATE = "Query error on host %s after %s ms: %s";
+
+    @VisibleForTesting
+    static final String TRUNCATED_OUTPUT = "... [truncated output]";
+
+    @VisibleForTesting
+    static final String FURTHER_PARAMS_OMITTED = " [further parameters omitted]";
+
+    @VisibleForTesting
+    static final String UNKNOWN_STATEMENT = "??Unknown Statement??";
+
+    private volatile long slowQueryLatencyThresholdMillis = DEFAULT_SLOW_QUERY_THRESHOLD_MS;
+
+    private volatile int maxQueryStringLength = DEFAULT_MAX_QUERY_STRING_LENGTH;
+
+    private volatile int maxParameterValueLength = DEFAULT_MAX_PARAMETER_VALUE_LENGTH;
+
+    private volatile int maxLoggedParameters = DEFAULT_MAX_LOGGED_PARAMETERS;
+
+    /**
+     * Private constructor. Instances of QueryLogger should be obtained via the {@link #builder(Cluster)} method.
+     */
+    private QueryLogger() {
+    }
+
+    /**
+     * Creates a new {@link QueryLogger.Builder} instance.
+     * <p>
+     * This is a convenience method for {@code new QueryLogger.Builder()}.
+     *
+     * @param cluster the {@link Cluster} this QueryLogger will be attached to.
+     *                Note: this parameter is currently not used, but it is included in the public API in anticipation of future use.
+     * @return the new QueryLogger builder.
+     */
+    public static QueryLogger.Builder builder(Cluster cluster) {
```

Comment:
As requested, the builder now takes a Cluster param in anticipation of future use.